Email Summary via Scheduled Jobs

undrash opened the pull request:
This is an example implementation for one-time scheduled jobs with Wasp.

There is an additional button titled "Schedule Summary Email" in the UI. If the user presses this button, they get a summary of yesterday's matches to their email.

![image](https://github.com/user-attachments/assets/50ddaee3-62d3-4bb2-afaa-aa66824b6155)

Below is a visualization of the email.
![image](https://github.com/user-attachments/assets/1bae4076-737b-4dc6-86b1-f63392d3b045)

infomiho commented on main.wasp:
```diff
@@ -13,6 +13,13 @@ app tennisScoreApp {
+      name: "Example",
+      email: "hello@itsme.com"
+    },
   }
```
Wasp apps using the `Dummy` sender can't be deployed, so we'd need to switch to something more production-worthy, maybe let's go with [Mailgun](https://www.mailgun.com) free plan in this case?

undrash replied:
Sounds good, implemented Mailgun with EU sending domain.

infomiho commented on example.env.server:
```diff
@@ -0,0 +1,3 @@
+DATABASE_URL=
+GOOGLE_CLIENT_ID=
+GOOGLE_CLIENT_SECRET=
```
Missing newline at the end of the file (git diffs are more clear if we include it)

undrash replied:
Done

infomiho commented on main.wasp:
```diff
@@ -67,3 +79,11 @@ route SignupRoute { path: "/signup", to: SignupPage }
+    fn: import { emailSummary } from "@src/workers/schedule",
+  },
+  entities: [Match]
+}
```
Missing newline at the end of the file (git diffs are more clear if we include it)

undrash replied:
Done

infomiho commented on main.wasp:
```diff
@@ -67,3 +79,11 @@ route SignupRoute { path: "/signup", to: SignupPage }
   component: import { SignupPage } from "@src/auth/SignupPage"
 }
+
+job scheduleSummaryEmailJob {
```
```suggestion
job sendEmailSummary {
```

This job does the actual sending, so no need to add `schedule` and also it's a job already, no need to suffix it with `job`.

undrash replied:
Refactored but left the Hungarian notation at the end so I am in line with the rest of the resources in that file. e.g. `tennisScoreApp` and `LoginRoute`

infomiho commented on src/components/Toast.tsx:
```diff
@@ -0,0 +1,89 @@
+      </button>
+    </div>
+  );
+} 
```
Missing newline at the end of the file (git diffs are more clear if we include it)

undrash replied:
Done

infomiho commented on src/matches/IndexPage.tsx:
```diff
@@ -10,10 +10,14 @@ import type { Match } from "./types";
   const { data: user } = useAuth();
+  const [showToast, setShowToast] = useState(false);
+  const [toastMessage, setToastMessage] = useState("");
+  const [toastType, setToastType] = useState<"success" | "error" | "info">("success");
```
Could we maybe go for an existing lib to use e.g. [sonner](https://www.npmjs.com/package/sonner) to avoid writing this logic ourselves? Let me know if you need help with figuring it out in Wasp (client root component should be the place to init the `<Toaster />` component).

infomiho commented on src/matches/IndexPage.tsx:
```diff
@@ -26,6 +30,22 @@ export function IndexPage() {
   const completedMatches = matches?.filter((match) => match.isComplete) || [];
 
+  const handleScheduleSummaryEmailClick = () => {
+    scheduleSummaryEmail({ sendAt: new Date().toISOString() })
```
This has very bad security implications, sending something from the client that the server accepts without validation. 

If the action is "schedule email for tomorrow", the server can easily do `new Date()` instead of the client doing it, that's what should we do :)

infomiho commented on src/matches/IndexPage.tsx:
```diff
@@ -57,6 +84,14 @@ export function IndexPage() {
         )}
       </div>
 
+      {/* Toast component */}
```
Redundant comment since it's obvious the component is right there (we might get rid of this anyways if we use the `sonner` lib instead).

infomiho commented on src/matches/operations.ts:
```diff
@@ -258,3 +261,14 @@ function formatMatchResponse(match: Match & { sets: Set[] }) {
+
+  await scheduleSummaryEmailJob.delay(sendAt).submit({ name: `Scheduled for ${sendAt}` });
+
+}) satisfies ScheduleSummaryEmail<{ sendAt: string }>;
```
Let's a newline here as well

infomiho commented on src/matches/validation.ts:
```diff
@@ -21,3 +21,7 @@ export const updateMatchVisibilitySchema = z.object({
   isPublic: z.boolean(),
 });
+
+export const scheduleEmailSummarySchema = z.object({
+  sendAt: z.string(),
+});
```
Good job on using the Zod schema to validate the payload. We might not need it in the end if the client doesn't send the date anymore in the end.

infomiho commented on src/utils/index.ts:
```diff
@@ -0,0 +1,115 @@
+  `;
+
+  return { textContent, htmlContent };
+}
```
Let's add a newline here as well

infomiho commented on src/utils/index.ts:
```diff
@@ -0,0 +1,115 @@
+import { type Match, type Set } from "wasp/entities";
+type MatchWithSets = Match & {
```
```suggestion
import { type Match, type Set } from "wasp/entities";

type MatchWithSets = Match & {
```

To keep it nicely formatted

infomiho commented on src/utils/index.ts:
```diff
@@ -0,0 +1,115 @@
+
+// Parse the tennis score into a readable format
+function formatTennisScore(match: MatchWithSets): string {
+  if (!match.score) return "No score available";
```
Do we even want to include those matches then if there is no score for the summary?

infomiho commented on src/utils/index.ts:
```diff
@@ -0,0 +1,115 @@
+    
+    const gamesScore = `${score.player1.games}-${score.player2.games}`;
+    
+    let setsInfo = "";
+    if (match.sets && Array.isArray(match.sets) && match.sets.length > 0) {
+      setsInfo = match.sets.map(set => 
+        `${set.player1Games}-${set.player2Games}`
+      ).join(", ");
+    }
```
You could convert this into a small helper fn that is called `getSetsInfo` which either returns an empty string or returns the result. This would enable you to avoid using `let` (which are a cause of many bugs).

undrash replied:
Done

infomiho commented on src/utils/index.ts:
```diff
@@ -0,0 +1,115 @@
+    
+    const status = match.isComplete ? "Completed" : "In progress";
+    
+    let scoreString = `${status}: `;
+    
+    if (setsInfo) {
+      scoreString += `Sets: [${setsInfo}]`;
+    }
+    
+    if (!match.isComplete) {
+      scoreString += ` Current game: ${currentGame}, Games: ${gamesScore}`;
+    }
+    
+    return scoreString;
```
Same thing here as above, you could go for a small helper fn like `getScoreString(status, setsInfo, match)` which returns the formatted string.

The string concatenation can be replaced with building an array that you `join` in the end (what you did above already).

undrash replied:
Done

infomiho commented on src/workers/schedule.ts:
```diff
@@ -0,0 +1,44 @@
+import { JSONObject } from "wasp/server/_types";
```
You can avoid this import and just use `{}` as the args type for the `ScheduleSummaryEmailJob` type.

undrash replied:
Tried that, thanks -- it worked. I am now using an `Input` type because I need the user `email` prop.

infomiho commented on src/workers/schedule.ts:
```diff
@@ -0,0 +1,44 @@
+
+export const emailSummary: ScheduleSummaryEmailJob<JSONObject, void> = async (_, context) => {
+  // Find yesterday's completed matches
+  const yesterday = new Date();
+  yesterday.setDate(yesterday.getDate() - 1);
+  yesterday.setHours(0, 0, 0, 0);
+  const today = new Date();
+  today.setHours(0, 0, 0, 0);
```
I'd avoid going the manual route and use something like [date-fns](https://date-fns.org/).

Claude says you can do the same thing like this (could be wrong):
```ts
import { subDays, startOfDay } from 'date-fns';

const today = startOfDay(new Date());
const yesterday = startOfDay(subDays(new Date(), 1));
```

undrash replied:
Sure, done

infomiho commented on src/workers/schedule.ts:
```diff
@@ -0,0 +1,44 @@
+      name: "Tennis Score App",
+      email: "notifications@tennisscoreapp.com",
+    },
+    to: "user@domain.com",
```
We need to use the actual user email. To me that signals that maybe the user email should be a part of the job's args. 

You use the `context.user` email in the action when you delay the job. https://wasp.sh/docs/auth/entities#authuser-object-fields 

Maybe we'll need to upgrade the app to save the email from the OAuth provider: https://wasp.sh/docs/auth/social-auth/google#using-the-data-received-from-google

undrash replied:
Sure thing, done.

infomiho reviewed (commented):
Great job navigating around Wasp. Left some comments, let me know in Discord if I can help you with anything.

undrash commented:
I'm wondering if something like this can be achieved within Wasp:
![image](https://github.com/user-attachments/assets/553a989e-d576-4b57-a36f-eda56ace7096)

undrash commented:
I'm wondering if we could use [Prettier](https://prettier.io/) for the project @infomiho? I haven't coded without an auto formatter in over 5 years so I'm struggling a bit :sweat_smile:

infomiho commented:
@undrash feel free to add a `.prettierrc` and use Pretter in the project 👍  You'll find the style we use in this PR's description https://github.com/wasp-lang/wasp/pull/2736

**On the topic of using `env` vars in the Wasp file** - not currently. It will be possible in the future when we switch to using Typescript for as our config language https://wasp.sh/docs/general/wasp-ts-config Hopefully soon 🤞

infomiho commented on migrations/20250515092430_useremail/migration.sql:
```diff
@@ -0,0 +1,12 @@
+  Warnings:
+
+  - A unique constraint covering the columns `[email]` on the table `User` will be added. If there are existing duplicate values, this will fail.
+  - Added the required column `email` to the `User` table without a default value. This is not possible if the table is not empty.
```
We'll have an issue here when we attempt to deploy this app in [production](https://tennis-score-app-client.fly.dev/) because we made the `email` required but maybe there are some existing users in the DB - which will make this migration fail.

We can either use an default value (but we don't have a sensible default value for an email) or make `email` optional. And we'll do one extra check in the application code "if the user has an email, you can schedule the summary, otherwise, you can't".

undrash replied:
Sure, done. Sorry I missed that, ironic given that I've used your migration article as reference for mine :stuck_out_tongue:

infomiho commented on schema.prisma:
```diff
@@ -9,6 +9,7 @@ generator client {
 
 model User {
   id      String  @id @default(cuid())
+  email   String  @unique
```
```suggestion
  email   String?  @unique
```

For context, read the comment on the migration file.

infomiho commented on src/auth/google.ts:
```diff
@@ -0,0 +1,11 @@
+import { defineUserSignupFields } from 'wasp/server/auth'
+
+export const userSignupFields = defineUserSignupFields({
+  email: (data: any) => data.profile.email,
```
Nitpick: since we already use `zod` for payload validation, we can use it here as well to avoid using `any`, here's an example in Open Saas how we do it for this exact thing: https://github.com/wasp-lang/open-saas/blob/main/template/app/src/auth/userSignupFields.ts#L72

undrash replied:
Done

infomiho commented on src/matches/IndexPage.tsx:
```diff
@@ -39,6 +52,13 @@ export function IndexPage() {
+            >
+              <Mail className="w-5 h-5 mr-2" />
+              Schedule Summary Email
+            </button>
```
Nitpick: I guess user can now click the button again while the existing request is still being sent, maybe we disable the button while the `scheduleSummaryEmail` action is loading?

undrash replied:
Sure, done

infomiho commented on src/matches/operations.ts:
```diff
@@ -258,3 +260,24 @@ function formatMatchResponse(match: Match & { sets: Set[] }) {
+    throw new HttpError(400, "User email not found");
+  }
+
+  // TODO: Update this date with the value you need (for example, tomorrow morning)
+  const sendAt = new Date().toISOString();
```
Is `sendAt` value used anywhere? I'd keep it simple and remove it since we don't use in the system.

undrash replied:
It is used in the `delay()` method below to demonstrate how to schedule a job (otherwise it's just sending instead of scheduling).

infomiho commented on src/matches/operations.ts:
```diff
@@ -258,3 +260,24 @@ function formatMatchResponse(match: Match & { sets: Set[] }) {
+  const sendAt = new Date().toISOString();
+
+  await sendEmailSummaryJob.delay(sendAt).submit({ 
+    name: `Scheduled for ${sendAt}`,
```
Is this `name` used anywhere, if not, I'd remove it as well.

infomiho commented on src/workers/schedule.ts:
```diff
@@ -0,0 +1,45 @@
+export const sendEmailSummary: SendEmailSummaryJob<Input, void> = async ({ email }, context) => {
+  // Find yesterday's completed matches
+  const today = startOfDay(new Date());
+  const yesterday = startOfDay(subDays(new Date(), 1));
```
```suggestion
  const today = startOfDay(new Date());
  const yesterday = startOfDay(subDays(today, 1));
```

undrash replied:
Done

undrash commented:
@infomiho Implemented the changes and I added prettier also with the config you linked me to, which led to some minor stylistic changes.

I also changed the single quotes to double quotes in JSX, because I saw a [discussion in the thread about it](https://github.com/wasp-lang/wasp/pull/2736/files#r2084880980).
